includes/class-cron-options-cpt.php

```diff
@@ -0,0 +1,375 @@
+
+class Cron_Options_CPT extends Singleton {
+	/**
+	 * PLUGIN SETUP
+	 */
+
+	/**
+	 * Class properties
+	 */
+	const LOCK = 'create-jobs';
+
+	const POST_TYPE   = 'a8c_cron_ctrl_events';
+	const POST_STATUS = 'inherit';
+
+	private $posts_to_clean = array();
+
+	private $option_before_unscheduling = null;
+
+	/**
+	 * Register hooks
+	 */
+	protected function class_init() {
+		// Data storage
+		add_action( 'init', array( $this, 'register_post_type' ) );
+
+		// Lock for post insertion, to guard against endless event creation when `wp_next_scheduled()` is misused
+		Lock::prime_lock( self::LOCK );
+
+		// Option interception
+		add_filter( 'pre_option_cron', array( $this, 'get_option' ) );
+		add_filter( 'pre_update_option_cron', array( $this, 'update_option' ), 10, 2 );
+	}
+
+	/**
+	 * Register a private post type to store cron events
+	 */
+	public function register_post_type() {
+		register_post_type( self::POST_TYPE, array(
+			'label'               => 'Cron Events',
+			'public'              => false,
+			'rewrite'             => false,
+			'export'              => false,
+			'exclude_from_search' => true,
+		) );
+
+		// Clear caches for any manually-inserted posts, lest stale caches be used
+		if ( ! empty( $this->posts_to_clean ) ) {
+			foreach ( $this->posts_to_clean as $index => $post_to_clean ) {
+				clean_post_cache( $post_to_clean );
+				unset( $this->posts_to_clean[ $index ] );
+			}
+		}
+	}
+
+	/**
+	 * PLUGIN FUNCTIONALITY
+	 */
+
+	/**
+	 * Override cron option requests with data from CPT
+	 */
+	public function get_option() {
+		$cron_array = array(
+			'version' => 2, // Core versions the cron array; without this, events will continually requeue
+		);
+
+		// Get events to re-render as the cron option
+		$page  = 1;
+
+		do {
+			$jobs_posts = $this->get_jobs( array(
+				'post_type'        => self::POST_TYPE,
+				'post_status'      => self::POST_STATUS,
+				'suppress_filters' => false,
+				'posts_per_page'   => 100,
+				'paged'            => $page,
+				'orderby'          => 'date',
+				'order'            => 'ASC',
+			) );
+
+			// Nothing more to add
+			if ( empty( $jobs_posts ) ) {
+				break;
+			}
+
+			$page++;
+
+			// Loop through results and built output Core expects
+			if ( ! empty( $jobs_posts ) ) {
+				foreach ( $jobs_posts as $jobs_post ) {
+					$timestamp = strtotime( $jobs_post->post_date_gmt );
+
+					$job_args = maybe_unserialize( $jobs_post->post_content_filtered );
+					if ( ! is_array( $job_args ) ) {
+						continue;
+					}
+
+					$action   = $job_args['action'];
+					$instance = $job_args['instance'];
+					$args     = $job_args['args'];
+
+					$cron_array[ $timestamp ][ $action ][ $instance ] = array(
+						'schedule' => $args['schedule'],
+						'args'     => $args['args'],
+					);
+
+					if ( isset( $args['interval'] ) ) {
+						$cron_array[ $timestamp ][ $action ][ $instance ]['interval'] = $args['interval'];
+					}
+
+				}
+			}
+		} while( true );
```

Comment:
We're probably safe to break after `$page > 10`? A site with 1000+ cron entries probably has bigger problems? My reasoning is mostly to protect ourselves if a logic error in the code fails to break out causing the loop to run forever and 💥 .